{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5b0a0e2",
   "metadata": {},
   "source": [
    "# Federated Protein Embeddings and Task Model Fitting with BioNeMo\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"> <b>NOTE</b> This notebook was tested on a single A1000 GPU and is compatible with BioNeMo Framework v2.2. To leverage additional or higher-performance GPUs, you can modify the configuration files and simulation script to accommodate multiple devices and increase thread utilization respectively. </div>\n",
    "\n",
    "This example notebook shows how to obtain protein learned representations in the form of embeddings using the ESM-1nv pre-trained model in a federated learning (FL) setting. The model is trained with NVIDIA's BioNeMo framework for Large Language Model training and inference. For more details, please visit NVIDIA BioNeMo Service at https://www.nvidia.com/en-us/gpu-cloud/bionemo.\n",
    "\n",
    "This example is based on NVIDIA BioNeMo Service [example](https://github.com/NVIDIA/BioNeMo/blob/main/examples/service/notebooks/task-fitting-predictor.ipynb) \n",
    "but runs inference locally (on the FL clients) instead of using BioNeMo's cloud API.\n",
    "\n",
    "This notebook will walk you through the task fitting workflow in the following sections:\n",
    "\n",
    "* Dataset sourcing & Data splitting\n",
    "* Federated embedding extraction\n",
    "* Training an MLP to predict subcellular location\n",
    "\n",
    "## Setup\n",
    "\n",
    "Ensure that you have read through the Getting Started section, can run the BioNeMo Framework docker container, and have configured the NGC Command Line Interface (CLI) within the container. It is assumed that this notebook is being executed from within the container.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"> <b>NOTE</b> Some of the cells below generate long text output.  We're using <pre>%%capture --no-display --no-stderr cell_output</pre> to suppress this output.  Comment or delete this line in the cells below to restore full output.</div>\n",
    "\n",
    "### Import and install all required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14dc1b1f-42ab-4a10-959a-3d14163fa974",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%capture --no-display --no-stderr cell_output\n",
    "! pip install nvflare~=2.5\n",
    "! pip install biopython\n",
    "! pip install scikit-learn\n",
    "! pip install matplotlib\n",
    "\n",
    "import io\n",
    "import numpy as np\n",
    "import os\n",
    "import pickle\n",
    "import re\n",
    "import requests\n",
    "import split_data\n",
    "\n",
    "from Bio import SeqIO\n",
    "from importlib import reload\n",
    "from nvflare import SimulatorRunner  \n",
    "from split_data import split"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c9c2167",
   "metadata": {},
   "source": [
    "### Obtaining the protein embeddings using the BioNeMo ESM-1nv model\n",
    "Using BioNeMo, each FL client can obtain numerical vector representations of protein sequences called embeddings. Protein embeddings can then be used for visualization or making downstream predictions.\n",
    "\n",
    "Here we are interested in training a neural network to predict subcellular location from an embedding.\n",
    "\n",
    "The data we will be using comes from the paper [Light attention predicts protein location from the language of life](https://academic.oup.com/bioinformaticsadvances/article/1/1/vbab035/6432029) by StÃ¤rk et al. In this paper, the authors developed a machine learning algorithm to predict the subcellular location of proteins from sequence through protein langage models that are similar to those hosted by BioNeMo. Protein subcellular location refers to where the protein localizes in the cell, for example a protein my be expressed in the Nucleus or in the Cytoplasm. Knowing where proteins localize can provide insights into the underlying mechanisms of cellular processes and help identify potential targets for drug development. The following image includes a few examples of subcellular locations in an animal cell:\n",
    "\n",
    "\n",
    "(Image freely available at https://pixabay.com/images/id-48542)\n",
    "\n",
    "### Dataset sourcing\n",
    "For our target input sequences, we will point to FASTA sequences in a benchmark dataset called Fitness Landscape Inference for Proteins (FLIP). FLIP encompasses experimental data across adeno-associated virus stability for gene therapy, protein domain B1 stability and immunoglobulin binding, and thermostability from multiple protein families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8407b137",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Example protein dataset location\n",
    "fasta_url= \"http://data.bioembeddings.com/public/FLIP/fasta/scl/mixed_soft.fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81fe63c2",
   "metadata": {},
   "source": [
    "First, we define the source of example protein dataset with the FASTA sequences. This data follows the [biotrainer](https://github.com/sacdallago/biotrainer/blob/main/docs/data_standardization.md) standard, so it includes information about the class in the FASTA header, and the protein sequence. Here are two example sequences in this file:\n",
    "\n",
    "```\n",
    ">Sequence1 TARGET=Cell_membrane SET=train VALIDATION=False\n",
    "MMKTLSSGNCTLNVPAKNSYRMVVLGASRVGKSSIVSRFLNGRFEDQYTPTIEDFHRKVYNIHGDMYQLDILDTSGNHPFPAM\n",
    "RRLSILTGDVFILVFSLDSRESFDEVKRLQKQILEVKSCLKNKTKEAAELPMVICGNKNDHSELCRQVPAMEAELLVSGDENC\n",
    "AYFEVSAKKNTNVNEMFYVLFSMAKLPHEMSPALHHKISVQYGDAFHPRPFCMRRTKVAGAYGMVSPFARRPSVNSDLKYIKA\n",
    "KVLREGQARERDKCSIQ\n",
    ">Sequence4833 TARGET=Nucleus SET=train VALIDATION=False\n",
    "MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRFRPGTVALREIRKYQKSTELLIRKLPFQRLVREIAQDFKTDL\n",
    "RFQSSAVAALQEAAEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
    "Note the following attributes in the FASTA header:\n",
    "```\n",
    "\n",
    "* `TARGET` attribute holds the subcellular location classification for the sequence, for instance Cell_membrane and Nucleus. This dataset includes a total of ten subcellelular location classes -- more on that below.\n",
    "* `SET` attribute defines whether the sequence should be used for training (train) or testing (test)\n",
    "* `VALIDATION` attribute defines whether the sequence should be used for validation (all sequences where this is True are also in set=train)\n",
    "\n",
    "### Downloading the protein sequences and subcellular location annotations\n",
    "In this step we download the FASTA file defined above and parse the sequences into a list of BioPython SeqRecord objects.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d7bfca4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Download the FASTA file from FLIP: https://github.com/J-SNACKKB/FLIP/tree/main/splits/scl\n",
    "fasta_content = requests.get(fasta_url, headers={\n",
    "    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x86)'\n",
    "}).content.decode('utf-8')\n",
    "fasta_stream = io.StringIO(fasta_content)\n",
    "\n",
    "# Obtain a list of SeqRecords/proteins which contain sequence and attributes\n",
    "# from the FASTA header\n",
    "proteins = list(SeqIO.parse(fasta_stream, \"fasta\"))\n",
    "print(f\"Downloaded {len(proteins)} sequences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e480628f",
   "metadata": {},
   "source": [
    "### Download Model Checkpoints\n",
    "\n",
    "The following code will download the pre-trained model, `esm2n/650m:2.0`, from the NGC registry:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4916a4c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from bionemo.core.data.load import load\n",
    "\n",
    "checkpoint_path = load(\"esm2/650m:2.0\", source=\"ngc\")\n",
    "print(checkpoint_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd955150",
   "metadata": {},
   "source": [
    "### Data splitting\n",
    "Next, we prepare the data for simulating federated learning using `n_clients`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "10fbc811",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 8619 valid sequences.\n",
      "Partition protein dataset with 10 classes into 3 sites with Dirichlet sampling under alpha 1.0\n",
      "{'site-1': {'Cell_membrane': 137,\n",
      "            'Cytoplasm': 311,\n",
      "            'Endoplasmic_reticulum': 123,\n",
      "            'Extracellular': 356,\n",
      "            'Golgi_apparatus': 43,\n",
      "            'Lysosome': 39,\n",
      "            'Mitochondrion': 250,\n",
      "            'Nucleus': 453,\n",
      "            'Peroxisome': 23,\n",
      "            'Plastid': 110},\n",
      " 'site-2': {'Cell_membrane': 217,\n",
      "            'Cytoplasm': 491,\n",
      "            'Endoplasmic_reticulum': 195,\n",
      "            'Extracellular': 562,\n",
      "            'Golgi_apparatus': 70,\n",
      "            'Lysosome': 63,\n",
      "            'Mitochondrion': 396,\n",
      "            'Nucleus': 716,\n",
      "            'Peroxisome': 38,\n",
      "            'Plastid': 174},\n",
      " 'site-3': {'Cell_membrane': 160,\n",
      "            'Cytoplasm': 361,\n",
      "            'Endoplasmic_reticulum': 144,\n",
      "            'Extracellular': 414,\n",
      "            'Golgi_apparatus': 51,\n",
      "            'Lysosome': 47,\n",
      "            'Mitochondrion': 292,\n",
      "            'Nucleus': 527,\n",
      "            'Peroxisome': 28,\n",
      "            'Plastid': 128}}\n",
      "Saved 1844 training and 1700 testing proteins for site-1, (10/10) unique train/test classes.\n",
      "Saved 2921 training and 1700 testing proteins for site-2, (10/10) unique train/test classes.\n",
      "Saved 2151 training and 1700 testing proteins for site-3, (10/10) unique train/test classes.\n",
      "Partition protein dataset with 10 classes into 3 sites with Dirichlet sampling under alpha 1.0\n",
      "{'site-1': {'Cell_membrane': 137,\n",
      "            'Cytoplasm': 311,\n",
      "            'Endoplasmic_reticulum': 123,\n",
      "            'Extracellular': 356,\n",
      "            'Golgi_apparatus': 43,\n",
      "            'Lysosome': 39,\n",
      "            'Mitochondrion': 250,\n",
      "            'Nucleus': 453,\n",
      "            'Peroxisome': 23,\n",
      "            'Plastid': 110},\n",
      " 'site-2': {'Cell_membrane': 217,\n",
      "            'Cytoplasm': 491,\n",
      "            'Endoplasmic_reticulum': 195,\n",
      "            'Extracellular': 562,\n",
      "            'Golgi_apparatus': 70,\n",
      "            'Lysosome': 63,\n",
      "            'Mitochondrion': 396,\n",
      "            'Nucleus': 716,\n",
      "            'Peroxisome': 38,\n",
      "            'Plastid': 174},\n",
      " 'site-3': {'Cell_membrane': 160,\n",
      "            'Cytoplasm': 361,\n",
      "            'Endoplasmic_reticulum': 144,\n",
      "            'Extracellular': 414,\n",
      "            'Golgi_apparatus': 51,\n",
      "            'Lysosome': 47,\n",
      "            'Mitochondrion': 292,\n",
      "            'Nucleus': 527,\n",
      "            'Peroxisome': 28,\n",
      "            'Plastid': 128}}\n",
      "Saved 1844 training and 1700 testing proteins for site-1, (10/10) unique train/test classes.\n",
      "Saved 2921 training and 1700 testing proteins for site-2, (10/10) unique train/test classes.\n",
      "Saved 2151 training and 1700 testing proteins for site-3, (10/10) unique train/test classes.\n"
     ]
    }
   ],
   "source": [
    "n_clients = 3\n",
    "# limiting to the proteins with sequence length<512 for embedding queries\n",
    "MAX_SEQUENCE_LEN = 512\n",
    "seed=0\n",
    "out_dir = \"/tmp/data/mixed_soft\"\n",
    "split_alpha = 1.0  # moderate label heterogeneity of alpha=1.0\n",
    "\n",
    "reload(split_data)\n",
    "\n",
    "np.random.seed(seed)\n",
    "\n",
    "# Extract meta data and split\n",
    "data = []\n",
    "for i, x in enumerate(proteins):\n",
    "        if len(str(x.seq)) > MAX_SEQUENCE_LEN:\n",
    "            continue\n",
    "            \n",
    "        entry = {key: value for key, value in re.findall(r\"([A-Z_]+)=(-?[A-z0-9]+[.0-9]*)\", x.description)}\n",
    "        entry[\"sequences\"] = str(x.seq)\n",
    "        entry[\"id\"] = str(i)\n",
    "        entry[\"labels\"] = entry[\"TARGET\"]\n",
    "       \n",
    "        data.append(entry)\n",
    "print(f\"Read {len(data)} valid sequences.\")\n",
    "               \n",
    "# Split the data and save for each client\n",
    "# Note, test_data is kept the same on each client and is not split\n",
    "# `concat=False` is used for SCL experiments (see ../downstream/scl)\n",
    "split(proteins=data, num_sites=n_clients, split_dir=out_dir, alpha=split_alpha, concat=False)  \n",
    "# `concat=True` is used for separate inference + MLP classifier in this notebook\n",
    "split(proteins=data, num_sites=n_clients, split_dir=out_dir, alpha=split_alpha, concat=True)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9fc9110",
   "metadata": {},
   "source": [
    "### Federated embedding extraction\n",
    "Running inference of the ESM-1nv model to extract embeddings requires a GPU with at least 12 GB memory. Here we run inference on each client sequentially using one thread to preserve GPU memory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72344e9b",
   "metadata": {},
   "source": [
    "First, copy the model into the job folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71854a35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!cp -r $checkpoint_path jobs/embeddings/app/models/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8073e435",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-12-04 03:08:37,653 - SimulatorRunner - INFO - Create the Simulator Server.\n",
      "2024-12-04 03:08:37,659 - CoreCell - INFO - server: creating listener on tcp://0:35267\n",
      "2024-12-04 03:08:37,673 - CoreCell - INFO - server: created backbone external listener for tcp://0:35267\n",
      "2024-12-04 03:08:37,674 - ConnectorManager - INFO - 35605: Try start_listener Listener resources: {'secure': False, 'host': 'localhost'}\n",
      "2024-12-04 03:08:37,676 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00002 PASSIVE tcp://0:2719] is starting\n",
      "2024-12-04 03:08:38,179 - CoreCell - INFO - server: created backbone internal listener for tcp://localhost:2719\n",
      "2024-12-04 03:08:38,180 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00001 PASSIVE tcp://0:35267] is starting\n",
      "2024-12-04 03:08:38,306 - nvflare.fuel.hci.server.hci - INFO - Starting Admin Server localhost on Port 41205\n",
      "2024-12-04 03:08:38,308 - SimulatorRunner - INFO - Deploy the Apps.\n",
      "2024-12-04 03:08:43,496 - SimulatorRunner - INFO - Create the simulate clients.\n",
      "2024-12-04 03:08:43,503 - ClientManager - INFO - Client: New client site-1@10.117.3.151 joined. Sent token: 5ab27765-5426-4825-9c1c-510e159720e9.  Total clients: 1\n",
      "2024-12-04 03:08:43,504 - FederatedClient - INFO - Successfully registered client:site-1 for project simulator_server. Token:5ab27765-5426-4825-9c1c-510e159720e9 SSID:\n",
      "2024-12-04 03:08:43,506 - ClientManager - INFO - Client: New client site-2@10.117.3.151 joined. Sent token: 3fc6f630-e87a-4cff-956d-a74ecf5a69f6.  Total clients: 2\n",
      "2024-12-04 03:08:43,508 - FederatedClient - INFO - Successfully registered client:site-2 for project simulator_server. Token:3fc6f630-e87a-4cff-956d-a74ecf5a69f6 SSID:\n",
      "2024-12-04 03:08:43,511 - ClientManager - INFO - Client: New client site-3@10.117.3.151 joined. Sent token: 9a895f17-5c3c-49a3-b5e9-5e29f485bd42.  Total clients: 3\n",
      "2024-12-04 03:08:43,512 - FederatedClient - INFO - Successfully registered client:site-3 for project simulator_server. Token:9a895f17-5c3c-49a3-b5e9-5e29f485bd42 SSID:\n",
      "2024-12-04 03:08:43,514 - SimulatorRunner - INFO - Set the client status ready.\n",
      "2024-12-04 03:08:43,515 - SimulatorRunner - INFO - Deploy and start the Server App.\n",
      "2024-12-04 03:08:43,518 - Cell - INFO - Register blob CB for channel='server_command', topic='*'\n",
      "2024-12-04 03:08:43,520 - Cell - INFO - Register blob CB for channel='aux_communication', topic='*'\n",
      "2024-12-04 03:08:43,521 - ServerCommandAgent - INFO - ServerCommandAgent cell register_request_cb: server.simulate_job\n",
      "2024-12-04 03:08:43,566 - AuxRunner - INFO - registered aux handler for topic __sync_runner__\n",
      "2024-12-04 03:08:43,568 - AuxRunner - INFO - registered aux handler for topic __job_heartbeat__\n",
      "2024-12-04 03:08:43,569 - AuxRunner - INFO - registered aux handler for topic __task_check__\n",
      "2024-12-04 03:08:43,570 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REQUEST\n",
      "2024-12-04 03:08:43,573 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REPLY\n",
      "2024-12-04 03:08:43,574 - ReliableMessage - INFO - enabled reliable message: max_request_workers=20 query_interval=2.0\n",
      "2024-12-04 03:08:43,576 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job]: Server runner starting ...\n",
      "2024-12-04 03:08:43,577 - AuxRunner - INFO - registered aux handler for topic fed.event\n",
      "2024-12-04 03:08:43,579 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job]: starting workflow bionemo_inference (<class 'bionemo_inference.BioNeMoInference'>) ...\n",
      "2024-12-04 03:08:43,580 - BioNeMoInference - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Initializing BroadcastAndProcess.\n",
      "2024-12-04 03:08:43,582 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Workflow bionemo_inference (<class 'bionemo_inference.BioNeMoInference'>) started\n",
      "2024-12-04 03:08:43,584 - WFCommServer - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: scheduled task bionemo_inference\n",
      "2024-12-04 03:08:44,518 - SimulatorClientRunner - INFO - Start the clients run simulation.\n",
      "2024-12-04 03:08:45,522 - SimulatorClientRunner - INFO - Simulate Run client: site-1 on GPU group: None\n",
      "2024-12-04 03:08:46,933 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00006 127.0.0.1:35267 <= 127.0.0.1:38946] is created: PID: 35605\n",
      "2024-12-04 03:08:46,840 - ClientTaskWorker - INFO - ClientTaskWorker started to run\n",
      "2024-12-04 03:08:46,930 - CoreCell - INFO - site-1.simulate_job: created backbone external connector to tcp://localhost:35267\n",
      "2024-12-04 03:08:46,930 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00001 ACTIVE tcp://localhost:35267] is starting\n",
      "2024-12-04 03:08:46,932 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 127.0.0.1:38946 => 127.0.0.1:35267] is created: PID: 35693\n",
      "2024-12-04 03:08:47,042 - AuxRunner - INFO - registered aux handler for topic __end_run__\n",
      "2024-12-04 03:08:47,042 - AuxRunner - INFO - registered aux handler for topic __do_task__\n",
      "2024-12-04 03:08:47,042 - Cell - INFO - Register blob CB for channel='aux_communication', topic='*'\n",
      "2024-12-04 03:08:47,563 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-1, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: assigned task to client site-1: name=bionemo_inference, id=4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:08:47,565 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-1, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: sent task assignment to client. client_name:site-1 task_id:4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:08:47,567 - GetTaskCommand - INFO - return task to client.  client_name: site-1  task_name: bionemo_inference   task_id: 4c3d722a-d699-4caa-83db-c089a5459629  sharable_header_task_id: 4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:08:47,548 - Cell - INFO - broadcast: channel='aux_communication', topic='__sync_runner__', targets=['server.simulate_job'], timeout=2.0\n",
      "2024-12-04 03:08:47,557 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: synced to Server Runner in 0.509739875793457 seconds\n",
      "2024-12-04 03:08:47,557 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REQUEST\n",
      "2024-12-04 03:08:47,557 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REPLY\n",
      "2024-12-04 03:08:47,558 - ReliableMessage - INFO - enabled reliable message: max_request_workers=20 query_interval=2.0\n",
      "2024-12-04 03:08:47,558 - AuxRunner - INFO - registered aux handler for topic fed.event\n",
      "2024-12-04 03:08:47,558 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: client runner started\n",
      "2024-12-04 03:08:47,559 - ClientTaskWorker - INFO - Initialize ClientRunner for client: site-1\n",
      "2024-12-04 03:08:47,571 - Communicator - INFO - Received from simulator_server server. getTask: bionemo_inference size: 686B (686 Bytes) time: 0.011802 seconds\n",
      "2024-12-04 03:08:47,571 - FederatedClient - INFO - pull_task completed. Task name:bionemo_inference Status:True \n",
      "2024-12-04 03:08:47,572 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job]: got task assignment: name=bionemo_inference, id=4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:08:47,572 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: invoking task executor BioNeMoLauncher\n",
      "2024-12-04 03:08:47,572 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Start Executor Launcher.\n",
      "2024-12-04 03:08:56,772 - SubprocessLauncher - INFO - 2024-12-04 03:08:56 - faiss.loader - INFO - Loading faiss with AVX2 support.\n",
      "2024-12-04 03:08:56,787 - SubprocessLauncher - INFO - 2024-12-04 03:08:56 - faiss.loader - INFO - Successfully loaded faiss with AVX2 support.\n",
      "2024-12-04 03:08:57,585 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:08:58,267 - SubprocessLauncher - INFO - GPU available: True (cuda), used: True\n",
      "2024-12-04 03:08:58,268 - SubprocessLauncher - INFO - TPU available: False, using: 0 TPU cores\n",
      "2024-12-04 03:08:58,268 - SubprocessLauncher - INFO - HPU available: False, using: 0 HPUs\n",
      "2024-12-04 03:08:58,337 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_strategy:310] Fixing mis-match between ddp-config & mcore-optimizer config\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:396] Rank 0 has data parallel group : [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:402] Rank 0 has combined group of data parallel and context parallel : [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:407] All data parallel group ranks with context parallel combined: [[0]]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:410] Ranks 0 has data parallel rank: 0\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:418] Rank 0 has context parallel group: [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:421] All context parallel group ranks: [[0]]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:422] Ranks 0 has context parallel rank: 0\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:429] Rank 0 has model parallel group: [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:430] All model parallel group ranks: [[0]]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:439] Rank 0 has tensor model parallel group: [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:443] All tensor model parallel group ranks: [[0]]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:444] Rank 0 has tensor model parallel rank: 0\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:464] Rank 0 has pipeline model parallel group: [0]\n",
      "2024-12-04 03:08:58,338 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:476] Rank 0 has embedding group: [0]\n",
      "2024-12-04 03:08:58,339 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:482] All pipeline model parallel group ranks: [[0]]\n",
      "2024-12-04 03:08:58,339 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:483] Rank 0 has pipeline model parallel rank 0\n",
      "2024-12-04 03:08:58,339 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:484] All embedding group ranks: [[0]]\n",
      "2024-12-04 03:08:58,339 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 megatron_init:485] Rank 0 has embedding rank: 0\n",
      "2024-12-04 03:08:58,339 - SubprocessLauncher - INFO - Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1\n",
      "2024-12-04 03:08:58,342 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:08:58,342 - SubprocessLauncher - INFO - distributed_backend=nccl\n",
      "2024-12-04 03:08:58,342 - SubprocessLauncher - INFO - All distributed processes registered. Starting with 1 processes\n",
      "2024-12-04 03:08:58,342 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:08:58,342 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:08:58,775 - SubprocessLauncher - INFO - WARNING: Logging before flag parsing goes to stderr.\n",
      "2024-12-04 03:08:58,776 - SubprocessLauncher - INFO - W1204 03:08:58.775758 139925838418368 config.py:85] Loading /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar\n",
      "2024-12-04 03:08:58,796 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:08:58 base:44] Padded vocab_size: 128, original vocab_size: 33, dummy tokens: 95.\n",
      "2024-12-04 03:08:59,277 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:08:59 nemo_logging:361] /usr/local/lib/python3.10/dist-packages/torch/distributed/checkpoint/state_dict_loader.py:25: UserWarning: 'load_state_dict' is deprecated and will be removed in future versions. Please use 'load' instead.\n",
      "2024-12-04 03:08:59,278 - SubprocessLauncher - INFO -       warnings.warn(\n",
      "2024-12-04 03:08:59,278 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:09:00,562 - SubprocessLauncher - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]\n",
      "2024-12-04 03:09:00,563 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:09:00 megatron_strategy:324] Could not copy Trainer's 'max_steps' to LR scheduler's 'max_steps'. If you are not using an LR scheduler, this warning can safely be ignored.\n",
      "2024-12-04 03:09:00,574 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:09:00 megatron_parallel:550]  > number of parameters on (tensor, pipeline) model parallel rank (0, 0): 651164288\n",
      "2024-12-04 03:09:07,596 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:09:17,606 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:09:27,617 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:09:37,627 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:09:47,638 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:09:57,648 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:07,659 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:17,669 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:27,680 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:37,690 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:47,701 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:10:57,711 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:11:07,721 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:11:17,732 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:11:25,524 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:11:25,524] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (8)\n",
      "2024-12-04 03:11:25,525 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:11:25,524] torch._dynamo.convert_frame: [WARNING]    function: 'forward' (/usr/local/lib/python3.10/dist-packages/bionemo/llm/model/layers.py:60)\n",
      "2024-12-04 03:11:25,525 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:11:25,524] torch._dynamo.convert_frame: [WARNING]    last reason: ___check_obj_id(L['self'], 139920203150576)                   # return query / self.sqrt_val  # bionemo/llm/model/layers.py:62 in forward\n",
      "2024-12-04 03:11:25,525 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:11:25,524] torch._dynamo.convert_frame: [WARNING] To log all recompilation reasons, use TORCH_LOGS=\"recompiles\".\n",
      "2024-12-04 03:11:25,525 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:11:25,524] torch._dynamo.convert_frame: [WARNING] To diagnose recompilation issues, see https://pytorch.org/docs/master/compile/troubleshooting.html.\n",
      "2024-12-04 03:11:27,742 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:11:29,211 - SubprocessLauncher - INFO - Writing output ['token_logits', 'input_ids', 'embeddings'] into inference_results.pt\n",
      "2024-12-04 03:11:37,765 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-1, peer_run=simulate_job]: got result from client site-1 for task: name=bionemo_inference, id=4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:11:37,769 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-1, peer_run=simulate_job, peer_rc=OK, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: finished processing client result by bionemo_inference\n",
      "2024-12-04 03:11:37,771 - SubmitUpdateCommand - INFO - submit_update process. client_name:site-1   task_id:4c3d722a-d699-4caa-83db-c089a5459629\n",
      "2024-12-04 03:11:37,776 - SimulatorClientRunner - INFO - Simulate Run client: site-2 on GPU group: None\n",
      "2024-12-04 03:11:37,781 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00006 Not Connected] is closed PID: 35605\n",
      "2024-12-04 03:11:37,753 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Run success\n",
      "2024-12-04 03:11:37,753 - BioNeMoLauncher - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: Stop Executor Launcher.\n",
      "2024-12-04 03:11:37,754 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: finished processing task\n",
      "2024-12-04 03:11:37,754 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: try #1: sending task result to server\n",
      "2024-12-04 03:11:37,755 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: checking task ...\n",
      "2024-12-04 03:11:37,755 - Cell - INFO - broadcast: channel='aux_communication', topic='__task_check__', targets=['server.simulate_job'], timeout=5.0\n",
      "2024-12-04 03:11:37,762 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: start to send task result to server\n",
      "2024-12-04 03:11:37,762 - FederatedClient - INFO - Starting to push execute result.\n",
      "2024-12-04 03:11:37,775 - Communicator - INFO -  SubmitUpdate size: 761B (761 Bytes). time: 0.012610 seconds\n",
      "2024-12-04 03:11:37,775 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=4c3d722a-d699-4caa-83db-c089a5459629]: task result sent to server\n",
      "2024-12-04 03:11:37,776 - ClientTaskWorker - INFO - Finished one task run for client: site-1 interval: 2 task_processed: True\n",
      "2024-12-04 03:11:37,779 - ClientTaskWorker - INFO - Clean up ClientRunner for : site-1 \n",
      "2024-12-04 03:11:37,781 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 Not Connected] is closed PID: 35693\n",
      "2024-12-04 03:11:40,194 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00007 127.0.0.1:35267 <= 127.0.0.1:42316] is created: PID: 35605\n",
      "2024-12-04 03:11:40,099 - ClientTaskWorker - INFO - ClientTaskWorker started to run\n",
      "2024-12-04 03:11:40,192 - CoreCell - INFO - site-2.simulate_job: created backbone external connector to tcp://localhost:35267\n",
      "2024-12-04 03:11:40,192 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00001 ACTIVE tcp://localhost:35267] is starting\n",
      "2024-12-04 03:11:40,193 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 127.0.0.1:42316 => 127.0.0.1:35267] is created: PID: 37687\n",
      "2024-12-04 03:11:40,304 - AuxRunner - INFO - registered aux handler for topic __end_run__\n",
      "2024-12-04 03:11:40,304 - AuxRunner - INFO - registered aux handler for topic __do_task__\n",
      "2024-12-04 03:11:40,304 - Cell - INFO - Register blob CB for channel='aux_communication', topic='*'\n",
      "2024-12-04 03:11:40,824 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-2, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: assigned task to client site-2: name=bionemo_inference, id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:11:40,826 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-2, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: sent task assignment to client. client_name:site-2 task_id:5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:11:40,828 - GetTaskCommand - INFO - return task to client.  client_name: site-2  task_name: bionemo_inference   task_id: 5b3ba7cf-f364-4f07-8c99-98a38e3e0b22  sharable_header_task_id: 5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:11:40,810 - Cell - INFO - broadcast: channel='aux_communication', topic='__sync_runner__', targets=['server.simulate_job'], timeout=2.0\n",
      "2024-12-04 03:11:40,818 - ClientRunner - INFO - [identity=site-2, run=simulate_job]: synced to Server Runner in 0.5086803436279297 seconds\n",
      "2024-12-04 03:11:40,818 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REQUEST\n",
      "2024-12-04 03:11:40,818 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REPLY\n",
      "2024-12-04 03:11:40,819 - ReliableMessage - INFO - enabled reliable message: max_request_workers=20 query_interval=2.0\n",
      "2024-12-04 03:11:40,819 - AuxRunner - INFO - registered aux handler for topic fed.event\n",
      "2024-12-04 03:11:40,820 - ClientRunner - INFO - [identity=site-2, run=simulate_job]: client runner started\n",
      "2024-12-04 03:11:40,820 - ClientTaskWorker - INFO - Initialize ClientRunner for client: site-2\n",
      "2024-12-04 03:11:40,832 - Communicator - INFO - Received from simulator_server server. getTask: bionemo_inference size: 686B (686 Bytes) time: 0.011977 seconds\n",
      "2024-12-04 03:11:40,833 - FederatedClient - INFO - pull_task completed. Task name:bionemo_inference Status:True \n",
      "2024-12-04 03:11:40,833 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job]: got task assignment: name=bionemo_inference, id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:11:40,833 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: invoking task executor BioNeMoLauncher\n",
      "2024-12-04 03:11:40,834 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Start Executor Launcher.\n",
      "2024-12-04 03:11:50,070 - SubprocessLauncher - INFO - 2024-12-04 03:11:50 - faiss.loader - INFO - Loading faiss with AVX2 support.\n",
      "2024-12-04 03:11:50,085 - SubprocessLauncher - INFO - 2024-12-04 03:11:50 - faiss.loader - INFO - Successfully loaded faiss with AVX2 support.\n",
      "2024-12-04 03:11:50,847 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:11:51,572 - SubprocessLauncher - INFO - GPU available: True (cuda), used: True\n",
      "2024-12-04 03:11:51,572 - SubprocessLauncher - INFO - TPU available: False, using: 0 TPU cores\n",
      "2024-12-04 03:11:51,572 - SubprocessLauncher - INFO - HPU available: False, using: 0 HPUs\n",
      "2024-12-04 03:11:51,641 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_strategy:310] Fixing mis-match between ddp-config & mcore-optimizer config\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:396] Rank 0 has data parallel group : [0]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:402] Rank 0 has combined group of data parallel and context parallel : [0]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:407] All data parallel group ranks with context parallel combined: [[0]]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:410] Ranks 0 has data parallel rank: 0\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:418] Rank 0 has context parallel group: [0]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:421] All context parallel group ranks: [[0]]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:422] Ranks 0 has context parallel rank: 0\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:429] Rank 0 has model parallel group: [0]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:430] All model parallel group ranks: [[0]]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:439] Rank 0 has tensor model parallel group: [0]\n",
      "2024-12-04 03:11:51,642 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:443] All tensor model parallel group ranks: [[0]]\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:444] Rank 0 has tensor model parallel rank: 0\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:464] Rank 0 has pipeline model parallel group: [0]\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:476] Rank 0 has embedding group: [0]\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:482] All pipeline model parallel group ranks: [[0]]\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:483] Rank 0 has pipeline model parallel rank 0\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:484] All embedding group ranks: [[0]]\n",
      "2024-12-04 03:11:51,643 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:51 megatron_init:485] Rank 0 has embedding rank: 0\n",
      "2024-12-04 03:11:51,644 - SubprocessLauncher - INFO - Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1\n",
      "2024-12-04 03:11:51,646 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:11:51,646 - SubprocessLauncher - INFO - distributed_backend=nccl\n",
      "2024-12-04 03:11:51,646 - SubprocessLauncher - INFO - All distributed processes registered. Starting with 1 processes\n",
      "2024-12-04 03:11:51,646 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:11:51,646 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:11:52,084 - SubprocessLauncher - INFO - WARNING: Logging before flag parsing goes to stderr.\n",
      "2024-12-04 03:11:52,084 - SubprocessLauncher - INFO - W1204 03:11:52.083841 140663632466368 config.py:85] Loading /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar\n",
      "2024-12-04 03:11:52,104 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:52 base:44] Padded vocab_size: 128, original vocab_size: 33, dummy tokens: 95.\n",
      "2024-12-04 03:11:52,581 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:11:52 nemo_logging:361] /usr/local/lib/python3.10/dist-packages/torch/distributed/checkpoint/state_dict_loader.py:25: UserWarning: 'load_state_dict' is deprecated and will be removed in future versions. Please use 'load' instead.\n",
      "2024-12-04 03:11:52,581 - SubprocessLauncher - INFO -       warnings.warn(\n",
      "2024-12-04 03:11:52,582 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:11:53,848 - SubprocessLauncher - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]\n",
      "2024-12-04 03:11:53,849 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:11:53 megatron_strategy:324] Could not copy Trainer's 'max_steps' to LR scheduler's 'max_steps'. If you are not using an LR scheduler, this warning can safely be ignored.\n",
      "2024-12-04 03:11:53,861 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:11:53 megatron_parallel:550]  > number of parameters on (tensor, pipeline) model parallel rank (0, 0): 651164288\n",
      "2024-12-04 03:12:00,857 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:12:10,868 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:12:20,878 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:12:30,889 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:12:40,899 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:12:50,909 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:00,920 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:10,930 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:20,941 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:30,951 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:40,962 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:13:50,972 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:14:00,983 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:14:10,993 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:14:19,314 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:14:19,313] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (8)\n",
      "2024-12-04 03:14:19,314 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:14:19,313] torch._dynamo.convert_frame: [WARNING]    function: 'forward' (/usr/local/lib/python3.10/dist-packages/bionemo/llm/model/layers.py:60)\n",
      "2024-12-04 03:14:19,314 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:14:19,313] torch._dynamo.convert_frame: [WARNING]    last reason: ___check_obj_id(L['self'], 140657993905392)                   # return query / self.sqrt_val  # bionemo/llm/model/layers.py:62 in forward\n",
      "2024-12-04 03:14:19,314 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:14:19,313] torch._dynamo.convert_frame: [WARNING] To log all recompilation reasons, use TORCH_LOGS=\"recompiles\".\n",
      "2024-12-04 03:14:19,314 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:14:19,313] torch._dynamo.convert_frame: [WARNING] To diagnose recompilation issues, see https://pytorch.org/docs/master/compile/troubleshooting.html.\n",
      "2024-12-04 03:14:21,001 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:14:23,063 - SubprocessLauncher - INFO - Writing output ['token_logits', 'input_ids', 'embeddings'] into inference_results.pt\n",
      "2024-12-04 03:14:31,025 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-2, peer_run=simulate_job]: got result from client site-2 for task: name=bionemo_inference, id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:14:31,028 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-2, peer_run=simulate_job, peer_rc=OK, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: finished processing client result by bionemo_inference\n",
      "2024-12-04 03:14:31,030 - SubmitUpdateCommand - INFO - submit_update process. client_name:site-2   task_id:5b3ba7cf-f364-4f07-8c99-98a38e3e0b22\n",
      "2024-12-04 03:14:31,035 - SimulatorClientRunner - INFO - Simulate Run client: site-3 on GPU group: None\n",
      "2024-12-04 03:14:31,039 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00007 Not Connected] is closed PID: 35605\n",
      "2024-12-04 03:14:31,012 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Run success\n",
      "2024-12-04 03:14:31,013 - BioNeMoLauncher - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: Stop Executor Launcher.\n",
      "2024-12-04 03:14:31,014 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: finished processing task\n",
      "2024-12-04 03:14:31,014 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: try #1: sending task result to server\n",
      "2024-12-04 03:14:31,014 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: checking task ...\n",
      "2024-12-04 03:14:31,015 - Cell - INFO - broadcast: channel='aux_communication', topic='__task_check__', targets=['server.simulate_job'], timeout=5.0\n",
      "2024-12-04 03:14:31,022 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: start to send task result to server\n",
      "2024-12-04 03:14:31,022 - FederatedClient - INFO - Starting to push execute result.\n",
      "2024-12-04 03:14:31,034 - Communicator - INFO -  SubmitUpdate size: 761B (761 Bytes). time: 0.011636 seconds\n",
      "2024-12-04 03:14:31,034 - ClientRunner - INFO - [identity=site-2, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=5b3ba7cf-f364-4f07-8c99-98a38e3e0b22]: task result sent to server\n",
      "2024-12-04 03:14:31,034 - ClientTaskWorker - INFO - Finished one task run for client: site-2 interval: 2 task_processed: True\n",
      "2024-12-04 03:14:31,037 - ClientTaskWorker - INFO - Clean up ClientRunner for : site-2 \n",
      "2024-12-04 03:14:31,039 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 Not Connected] is closed PID: 37687\n",
      "2024-12-04 03:14:33,455 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00008 127.0.0.1:35267 <= 127.0.0.1:45174] is created: PID: 35605\n",
      "2024-12-04 03:14:33,359 - ClientTaskWorker - INFO - ClientTaskWorker started to run\n",
      "2024-12-04 03:14:33,452 - CoreCell - INFO - site-3.simulate_job: created backbone external connector to tcp://localhost:35267\n",
      "2024-12-04 03:14:33,452 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00001 ACTIVE tcp://localhost:35267] is starting\n",
      "2024-12-04 03:14:33,454 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 127.0.0.1:45174 => 127.0.0.1:35267] is created: PID: 39678\n",
      "2024-12-04 03:14:33,564 - AuxRunner - INFO - registered aux handler for topic __end_run__\n",
      "2024-12-04 03:14:33,564 - AuxRunner - INFO - registered aux handler for topic __do_task__\n",
      "2024-12-04 03:14:33,565 - Cell - INFO - Register blob CB for channel='aux_communication', topic='*'\n",
      "2024-12-04 03:14:34,084 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-3, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: assigned task to client site-3: name=bionemo_inference, id=1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:14:34,086 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-3, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: sent task assignment to client. client_name:site-3 task_id:1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:14:34,088 - GetTaskCommand - INFO - return task to client.  client_name: site-3  task_name: bionemo_inference   task_id: 1f275896-a2f8-4204-8a6d-9c395ec5c21c  sharable_header_task_id: 1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:14:34,070 - Cell - INFO - broadcast: channel='aux_communication', topic='__sync_runner__', targets=['server.simulate_job'], timeout=2.0\n",
      "2024-12-04 03:14:34,078 - ClientRunner - INFO - [identity=site-3, run=simulate_job]: synced to Server Runner in 0.5083255767822266 seconds\n",
      "2024-12-04 03:14:34,078 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REQUEST\n",
      "2024-12-04 03:14:34,078 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REPLY\n",
      "2024-12-04 03:14:34,079 - ReliableMessage - INFO - enabled reliable message: max_request_workers=20 query_interval=2.0\n",
      "2024-12-04 03:14:34,079 - AuxRunner - INFO - registered aux handler for topic fed.event\n",
      "2024-12-04 03:14:34,080 - ClientRunner - INFO - [identity=site-3, run=simulate_job]: client runner started\n",
      "2024-12-04 03:14:34,080 - ClientTaskWorker - INFO - Initialize ClientRunner for client: site-3\n",
      "2024-12-04 03:14:34,092 - Communicator - INFO - Received from simulator_server server. getTask: bionemo_inference size: 686B (686 Bytes) time: 0.012058 seconds\n",
      "2024-12-04 03:14:34,093 - FederatedClient - INFO - pull_task completed. Task name:bionemo_inference Status:True \n",
      "2024-12-04 03:14:34,093 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job]: got task assignment: name=bionemo_inference, id=1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:14:34,094 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: invoking task executor BioNeMoLauncher\n",
      "2024-12-04 03:14:34,094 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Start Executor Launcher.\n",
      "2024-12-04 03:14:43,342 - SubprocessLauncher - INFO - 2024-12-04 03:14:43 - faiss.loader - INFO - Loading faiss with AVX2 support.\n",
      "2024-12-04 03:14:43,357 - SubprocessLauncher - INFO - 2024-12-04 03:14:43 - faiss.loader - INFO - Successfully loaded faiss with AVX2 support.\n",
      "2024-12-04 03:14:44,107 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:14:44,836 - SubprocessLauncher - INFO - GPU available: True (cuda), used: True\n",
      "2024-12-04 03:14:44,836 - SubprocessLauncher - INFO - TPU available: False, using: 0 TPU cores\n",
      "2024-12-04 03:14:44,836 - SubprocessLauncher - INFO - HPU available: False, using: 0 HPUs\n",
      "2024-12-04 03:14:44,905 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_strategy:310] Fixing mis-match between ddp-config & mcore-optimizer config\n",
      "2024-12-04 03:14:44,905 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:396] Rank 0 has data parallel group : [0]\n",
      "2024-12-04 03:14:44,906 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:402] Rank 0 has combined group of data parallel and context parallel : [0]\n",
      "2024-12-04 03:14:44,906 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:407] All data parallel group ranks with context parallel combined: [[0]]\n",
      "2024-12-04 03:14:44,906 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:410] Ranks 0 has data parallel rank: 0\n",
      "2024-12-04 03:14:44,906 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:418] Rank 0 has context parallel group: [0]\n",
      "2024-12-04 03:14:44,906 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:421] All context parallel group ranks: [[0]]\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:422] Ranks 0 has context parallel rank: 0\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:429] Rank 0 has model parallel group: [0]\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:430] All model parallel group ranks: [[0]]\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:439] Rank 0 has tensor model parallel group: [0]\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:443] All tensor model parallel group ranks: [[0]]\n",
      "2024-12-04 03:14:44,907 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:444] Rank 0 has tensor model parallel rank: 0\n",
      "2024-12-04 03:14:44,908 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:464] Rank 0 has pipeline model parallel group: [0]\n",
      "2024-12-04 03:14:44,908 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:476] Rank 0 has embedding group: [0]\n",
      "2024-12-04 03:14:44,908 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:482] All pipeline model parallel group ranks: [[0]]\n",
      "2024-12-04 03:14:44,908 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:483] Rank 0 has pipeline model parallel rank 0\n",
      "2024-12-04 03:14:44,908 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:484] All embedding group ranks: [[0]]\n",
      "2024-12-04 03:14:44,909 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:44 megatron_init:485] Rank 0 has embedding rank: 0\n",
      "2024-12-04 03:14:44,909 - SubprocessLauncher - INFO - Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1\n",
      "2024-12-04 03:14:44,912 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:14:44,912 - SubprocessLauncher - INFO - distributed_backend=nccl\n",
      "2024-12-04 03:14:44,912 - SubprocessLauncher - INFO - All distributed processes registered. Starting with 1 processes\n",
      "2024-12-04 03:14:44,912 - SubprocessLauncher - INFO - ----------------------------------------------------------------------------------------------------\n",
      "2024-12-04 03:14:44,912 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:14:45,354 - SubprocessLauncher - INFO - WARNING: Logging before flag parsing goes to stderr.\n",
      "2024-12-04 03:14:45,354 - SubprocessLauncher - INFO - W1204 03:14:45.354014 140159471206848 config.py:85] Loading /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar\n",
      "2024-12-04 03:14:45,375 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:45 base:44] Padded vocab_size: 128, original vocab_size: 33, dummy tokens: 95.\n",
      "2024-12-04 03:14:45,859 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:14:45 nemo_logging:361] /usr/local/lib/python3.10/dist-packages/torch/distributed/checkpoint/state_dict_loader.py:25: UserWarning: 'load_state_dict' is deprecated and will be removed in future versions. Please use 'load' instead.\n",
      "2024-12-04 03:14:45,860 - SubprocessLauncher - INFO -       warnings.warn(\n",
      "2024-12-04 03:14:45,860 - SubprocessLauncher - INFO - \n",
      "2024-12-04 03:14:47,139 - SubprocessLauncher - INFO - LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]\n",
      "2024-12-04 03:14:47,140 - SubprocessLauncher - INFO - [NeMo W 2024-12-04 03:14:47 megatron_strategy:324] Could not copy Trainer's 'max_steps' to LR scheduler's 'max_steps'. If you are not using an LR scheduler, this warning can safely be ignored.\n",
      "2024-12-04 03:14:47,151 - SubprocessLauncher - INFO - [NeMo I 2024-12-04 03:14:47 megatron_parallel:550]  > number of parameters on (tensor, pipeline) model parallel rank (0, 0): 651164288\n",
      "2024-12-04 03:14:54,117 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:04,128 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:14,129 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:24,140 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:34,150 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:44,160 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:15:54,171 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:04,177 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:14,188 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:24,189 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:34,200 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:44,210 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:16:54,220 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:17:04,231 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:17:11,989 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:17:11,989] torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (8)\n",
      "2024-12-04 03:17:11,989 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:17:11,989] torch._dynamo.convert_frame: [WARNING]    function: 'forward' (/usr/local/lib/python3.10/dist-packages/bionemo/llm/model/layers.py:60)\n",
      "2024-12-04 03:17:11,989 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:17:11,989] torch._dynamo.convert_frame: [WARNING]    last reason: ___check_obj_id(L['self'], 140153834271984)                   # return query / self.sqrt_val  # bionemo/llm/model/layers.py:62 in forward\n",
      "2024-12-04 03:17:11,990 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:17:11,989] torch._dynamo.convert_frame: [WARNING] To log all recompilation reasons, use TORCH_LOGS=\"recompiles\".\n",
      "2024-12-04 03:17:11,990 - SubprocessLauncher - INFO - [rank0]:[2024-12-04 03:17:11,989] torch._dynamo.convert_frame: [WARNING] To diagnose recompilation issues, see https://pytorch.org/docs/master/compile/troubleshooting.html.\n",
      "2024-12-04 03:17:14,241 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Check running command: infer_esm2 --checkpoint-path /root/.cache/bionemo/0798767e843e3d54315aef91934d28ae7d8e93c2849d5fcfbdf5fac242013997-esm2_650M_nemo2.tar.gz.untar --data-path /tmp/data/mixed_soft/data_site-1.csv --results-path ./inference_results.pt --precision fp32 --include-embeddings --include-logits --include-input-ids\n",
      "2024-12-04 03:17:15,678 - SubprocessLauncher - INFO - Writing output ['token_logits', 'input_ids', 'embeddings'] into inference_results.pt\n",
      "2024-12-04 03:17:24,264 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-3, peer_run=simulate_job]: got result from client site-3 for task: name=bionemo_inference, id=1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:17:24,268 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-3, peer_run=simulate_job, peer_rc=OK, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: finished processing client result by bionemo_inference\n",
      "2024-12-04 03:17:24,270 - SubmitUpdateCommand - INFO - submit_update process. client_name:site-3   task_id:1f275896-a2f8-4204-8a6d-9c395ec5c21c\n",
      "2024-12-04 03:17:24,272 - WFCommServer - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: task bionemo_inference exit with status TaskCompletionStatus.OK\n",
      "2024-12-04 03:17:24,276 - SimulatorClientRunner - INFO - Simulate Run client: site-1 on GPU group: None\n",
      "2024-12-04 03:17:24,281 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00008 Not Connected] is closed PID: 35605\n",
      "2024-12-04 03:17:24,395 - BioNeMoInferenceProcessor - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Clients inference responses:\n",
      "{'site-1': {'bionemo_number_sequences': 1},\n",
      " 'site-2': {'bionemo_number_sequences': 1},\n",
      " 'site-3': {'bionemo_number_sequences': 1}}\n",
      "2024-12-04 03:17:24,398 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Workflow: bionemo_inference finalizing ...\n",
      "2024-12-04 03:17:24,400 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: ABOUT_TO_END_RUN fired\n",
      "2024-12-04 03:17:24,406 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Firing CHECK_END_RUN_READINESS ...\n",
      "2024-12-04 03:17:24,408 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: END_RUN fired\n",
      "2024-12-04 03:17:24,410 - ReliableMessage - INFO - ReliableMessage is shutdown\n",
      "2024-12-04 03:17:24,412 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: Server runner finished.\n",
      "2024-12-04 03:17:24,252 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Run success\n",
      "2024-12-04 03:17:24,253 - BioNeMoLauncher - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: Stop Executor Launcher.\n",
      "2024-12-04 03:17:24,253 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: finished processing task\n",
      "2024-12-04 03:17:24,254 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: try #1: sending task result to server\n",
      "2024-12-04 03:17:24,254 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: checking task ...\n",
      "2024-12-04 03:17:24,254 - Cell - INFO - broadcast: channel='aux_communication', topic='__task_check__', targets=['server.simulate_job'], timeout=5.0\n",
      "2024-12-04 03:17:24,261 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: start to send task result to server\n",
      "2024-12-04 03:17:24,261 - FederatedClient - INFO - Starting to push execute result.\n",
      "2024-12-04 03:17:24,275 - Communicator - INFO -  SubmitUpdate size: 761B (761 Bytes). time: 0.013560 seconds\n",
      "2024-12-04 03:17:24,275 - ClientRunner - INFO - [identity=site-3, run=simulate_job, peer=simulator_server, peer_run=simulate_job, task_name=bionemo_inference, task_id=1f275896-a2f8-4204-8a6d-9c395ec5c21c]: task result sent to server\n",
      "2024-12-04 03:17:24,276 - ClientTaskWorker - INFO - Finished one task run for client: site-3 interval: 2 task_processed: True\n",
      "2024-12-04 03:17:24,279 - ClientTaskWorker - INFO - Clean up ClientRunner for : site-3 \n",
      "2024-12-04 03:17:24,281 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 Not Connected] is closed PID: 39678\n",
      "2024-12-04 03:17:26,174 - SimulatorServer - INFO - Server app stopped.\n",
      "\n",
      "\n",
      "2024-12-04 03:17:26,197 - ReliableMessage - INFO - shutdown reliable message monitor\n",
      "2024-12-04 03:17:26,425 - nvflare.fuel.hci.server.hci - INFO - Admin Server localhost on Port 41205 shutdown!\n",
      "2024-12-04 03:17:26,427 - SimulatorServer - INFO - shutting down server\n",
      "2024-12-04 03:17:26,429 - SimulatorServer - INFO - canceling sync locks\n",
      "2024-12-04 03:17:26,430 - SimulatorServer - INFO - server off\n",
      "2024-12-04 03:17:26,681 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00009 127.0.0.1:35267 <= 127.0.0.1:55458] is created: PID: 35605\n",
      "2024-12-04 03:17:26,603 - ClientTaskWorker - INFO - ClientTaskWorker started to run\n",
      "2024-12-04 03:17:26,679 - CoreCell - INFO - site-1.simulate_job: created backbone external connector to tcp://localhost:35267\n",
      "2024-12-04 03:17:26,679 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connector [CH00001 ACTIVE tcp://localhost:35267] is starting\n",
      "2024-12-04 03:17:26,680 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 127.0.0.1:55458 => 127.0.0.1:35267] is created: PID: 41686\n",
      "2024-12-04 03:17:26,789 - AuxRunner - INFO - registered aux handler for topic __end_run__\n",
      "2024-12-04 03:17:26,789 - AuxRunner - INFO - registered aux handler for topic __do_task__\n",
      "2024-12-04 03:17:26,790 - Cell - INFO - Register blob CB for channel='aux_communication', topic='*'\n",
      "2024-12-04 03:17:27,308 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference, peer=site-1, peer_run=simulate_job]: server runner is finalizing - asked client to end the run\n",
      "2024-12-04 03:17:27,311 - GetTaskCommand - INFO - return task to client.  client_name: site-1  task_name: __end_run__   task_id:   sharable_header_task_id: \n",
      "2024-12-04 03:17:27,295 - Cell - INFO - broadcast: channel='aux_communication', topic='__sync_runner__', targets=['server.simulate_job'], timeout=2.0\n",
      "2024-12-04 03:17:27,303 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: synced to Server Runner in 0.5080511569976807 seconds\n",
      "2024-12-04 03:17:27,303 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REQUEST\n",
      "2024-12-04 03:17:27,303 - AuxRunner - INFO - registered aux handler for topic RM.RELIABLE_REPLY\n",
      "2024-12-04 03:17:27,303 - ReliableMessage - INFO - enabled reliable message: max_request_workers=20 query_interval=2.0\n",
      "2024-12-04 03:17:27,304 - AuxRunner - INFO - registered aux handler for topic fed.event\n",
      "2024-12-04 03:17:27,304 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: client runner started\n",
      "2024-12-04 03:17:27,304 - ClientTaskWorker - INFO - Initialize ClientRunner for client: site-1\n",
      "2024-12-04 03:17:27,314 - FederatedClient - INFO - pull_task completed. Task name:__end_run__ Status:True \n",
      "2024-12-04 03:17:27,315 - ClientRunner - INFO - [identity=site-1, run=simulate_job, peer=simulator_server, peer_run=simulate_job]: server asked to end the run\n",
      "2024-12-04 03:17:27,315 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: started end-run events sequence\n",
      "2024-12-04 03:17:27,315 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: ABOUT_TO_END_RUN fired\n",
      "2024-12-04 03:17:27,315 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: Firing CHECK_END_RUN_READINESS ...\n",
      "2024-12-04 03:17:27,816 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: Firing CHECK_END_RUN_READINESS ...\n",
      "2024-12-04 03:17:28,317 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: Firing CHECK_END_RUN_READINESS ...\n",
      "2024-12-04 03:17:28,861 - SimulatorClientRunner - INFO - Simulate Run client: site-2 on GPU group: None\n",
      "2024-12-04 03:17:28,865 - FederatedClient - INFO - Shutting down client run: site-1\n",
      "2024-12-04 03:17:28,867 - FederatedClient - INFO - Shutting down client run: site-2\n",
      "2024-12-04 03:17:28,867 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00009 Not Connected] is closed PID: 35605\n",
      "2024-12-04 03:17:28,869 - FederatedClient - INFO - Shutting down client run: site-3\n",
      "2024-12-04 03:17:28,871 - ServerRunner - INFO - [identity=simulator_server, run=simulate_job, wf=bionemo_inference]: asked to abort - triggered abort_signal to stop the RUN\n",
      "2024-12-04 03:17:28,819 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: Firing CHECK_END_RUN_READINESS ...\n",
      "2024-12-04 03:17:28,819 - ClientRunner - INFO - [identity=site-1, run=simulate_job]: END_RUN fired\n",
      "2024-12-04 03:17:28,819 - ClientTaskWorker - INFO - End the Simulator run.\n",
      "2024-12-04 03:17:28,864 - ClientTaskWorker - INFO - Clean up ClientRunner for : site-1 \n",
      "2024-12-04 03:17:28,867 - nvflare.fuel.f3.sfm.conn_manager - INFO - Connection [CN00002 Not Connected] is closed PID: 41686\n",
      "2024-12-04 03:17:32,420 - MPM - INFO - MPM: Good Bye!\n",
      "Simulator finished with run_status 0\n"
     ]
    }
   ],
   "source": [
    "simulator = SimulatorRunner(\n",
    "    job_folder=\"jobs/embeddings\",\n",
    "    workspace=\"/tmp/nvflare/bionemo/embeddings\",\n",
    "    n_clients=n_clients,\n",
    "    threads=1  # due to memory constraints, we run the client execution sequentially in one thread\n",
    ")\n",
    "run_status = simulator.run()\n",
    "print(\"Simulator finished with run_status\", run_status)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6448e76d",
   "metadata": {},
   "source": [
    "### Inspecting the embeddings and labels\n",
    "Embeddings returned from the BioNeMo model are vectors of fixed size for each input sequence. In other words, if we input 10 sequences, we will obtain a matrix `10xD`, where `D` is the size of the embedding (in the case of ESM2-650m, `D=1280`). At a glance, these real-valued vector embeddings don't show any obvious features (see the printout in the next cell). But these vectors do contain information that can be used in downstream models to reveal properties of the protein, for example the subcellular location as we'll explore below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "62cb283d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "token_logits\ttorch.Size([1024, 3543, 128])\n",
      "input_ids\ttorch.Size([3543, 1024])\n",
      "embeddings\ttorch.Size([3543, 1280])\n",
      "Loaded 3543 embeddings from site-1.\n",
      "1. embedding: range -5.90-0.35, mean=-0.00, shape=torch.Size([1280])\n",
      "2. embedding: range -5.75-0.30, mean=-0.00, shape=torch.Size([1280])\n",
      "3. embedding: range -6.08-0.31, mean=-0.00, shape=torch.Size([1280])\n",
      "4. embedding: range -7.75-0.65, mean=-0.00, shape=torch.Size([1280])\n"
     ]
    }
   ],
   "source": [
    "# load embeddings from site-1\n",
    "import torch\n",
    "results = torch.load(\"/tmp/nvflare/bionemo/embeddings/site-1/simulate_job/app_site-1/inference_results.pt\")\n",
    "\n",
    "for key, val in results.items():\n",
    "    if val is not None:\n",
    "        print(f'{key}\\t{val.shape}')\n",
    "\n",
    "protein_embeddings = results['embeddings']\n",
    "print(f\"Loaded {len(protein_embeddings)} embeddings from site-1.\")\n",
    "\n",
    "for i in range(4):\n",
    "    x = protein_embeddings[i]\n",
    "    print(f\"{i+1}. embedding: range {x.min():.2f}-{x.max():.2f}, mean={x.mean():.2f}, shape={x.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb1bb7dc",
   "metadata": {},
   "source": [
    "Let's enumerate the labels corresponding to potential subcellular locations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e6d0cd1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. Cytoplasm\n",
      "2. Plastid\n",
      "3. Cell membrane\n",
      "4. Endoplasmic reticulum\n",
      "5. Extracellular\n",
      "6. Nucleus\n",
      "7. Golgi apparatus\n",
      "8. Lysosome\n",
      "9. Peroxisome\n",
      "10. Mitochondrion\n"
     ]
    }
   ],
   "source": [
    "# Let's also print all the labels\n",
    "\n",
    "labels = set([entry['labels'] for entry in data])\n",
    "\n",
    "for i, label in enumerate(labels):\n",
    "    print(f\"{i+1}. {label.replace('_', ' ')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8d9eefb",
   "metadata": {},
   "source": [
    "### Training a MLP to predict subcellular location\n",
    "To be able to classify proteins for their subcellular location, we train a simple scikit-learn Multi-layer Perceptron (MPL) classifier using Federated Averaging ([FedAvg](https://arxiv.org/abs/1602.05629)). The MLP model uses a network of hidden layers to fit the input embedding vectors to the model classes (the cellular locations above). In the simulation below, we define the MLP to use the Adam optimizer with a network of (512, 256, 128) hidden layers, defining a random state (or seed) for reproducibility, and trained for 30 rounds of FedAvg (see [config_fed_server.json](./jobs/fedavg/app/config/config_fed_server.json)). \n",
    "\n",
    "We can use the same configuration also to simulate local training where each client is only training with their own data by setting `os.environ[\"SIM_LOCAL\"] = \"True\"`. Our [BioNeMoMLPLearner](./jobs/fedavg/app/custom/bionemo_mlp_learner.py) will then ignore the global weights coming from the server.\n",
    "\n",
    "### Local training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72a9dad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "os.environ[\"SIM_LOCAL\"] = \"True\"\n",
    "\n",
    "simulator = SimulatorRunner(\n",
    "    job_folder=\"jobs/fedavg\",\n",
    "    workspace=f\"/tmp/nvflare/bionemo/local_alpha{split_alpha}\",\n",
    "    n_clients=n_clients,\n",
    "    threads=n_clients\n",
    ")\n",
    "run_status = simulator.run()\n",
    "print(\"Simulator finished with run_status\", run_status)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b58f1f5",
   "metadata": {},
   "source": [
    "### Federated learning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67275cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"SIM_LOCAL\"] = \"False\"\n",
    "\n",
    "simulator = SimulatorRunner(\n",
    "    job_folder=\"jobs/fedavg\",\n",
    "    workspace=f\"/tmp/nvflare/bionemo/fedavg_alpha{split_alpha}\",\n",
    "    n_clients=n_clients,\n",
    "    threads=n_clients\n",
    ")\n",
    "run_status = simulator.run()\n",
    "print(\"Simulator finished with run_status\", run_status)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "791d6304",
   "metadata": {},
   "source": [
    "### TensorBoard Visualization\n",
    "You can visualize the training progress using TensorBoard\n",
    "```\n",
    "tensorboard --logdir /tmp/nvflare/bionemo\n",
    "```\n",
    "\n",
    "An example of local (red) vs federated (blue) training is shown below.\n",
    "\n",
    "![TensorBoard training curves](tb_curve.png)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
